external/fv3fit/fv3fit/emulation/layers/architecture.py

```diff
@@ -178,3 +261,165 @@ def NoWeightSharingSLP(
+    vertical dimension.
+
+    E.g., Combined inputs [nsamples, 17, 4] -> RNN network out [nsamples, 17, channels]
+        -> connected output [nsamples, 17]
+
+    If an output is a single-level (i.e., feature length of 1)
+    then it is connected to the lowest layer containing the full vertical information.
+    """
+
+    def __init__(self, feature_lengths: Mapping[str, int], *args, **kwargs):
+        """
+        Args:
+            feature_lengths: Map of output variable names to expected
+                feature dimension length
+        """
+        super().__init__(*args, **kwargs)
+        self._feature_lengths = feature_lengths
+
+        self.output_layers = {
+            name: tf.keras.layers.LocallyConnected1D(1, 1, name=f"rnn_output_{name}")
+            for name in self._feature_lengths.keys()
+        }
+
+    def call(self, rnn_outputs: tf.Tensor) -> Mapping[str, tf.Tensor]:
+        """
+        Args:
+            hidden_output: Network output from hidden layers of RNN with dim
+                [nsamples, nlev, channels]
+
+        Returns:
+            Connected model outputs by variable
+        """
+
+        fields = {}
+        for name, feature_length in self._feature_lengths.items():
+            if feature_length == 1:
+                rnn_out = rnn_outputs[..., 0:1, :]
+            else:
+                rnn_out = rnn_outputs
+
+            field_out = self.output_layers[name](rnn_out)
+            field_out = tf.squeeze(field_out, axis=-1)
+            fields[name] = field_out
+
+        return fields
+
+
+def _get_output_layer(key, feature_lengths):
+    if key == "rnn-v1":
+        return RNNOutput(feature_lengths)
+    else:
+        return StandardOutput(feature_lengths)
+
+
+def _get_combine_layer(key):
+    if key == "rnn-v1" or key == "rnn":
+        return CombineInputs(combine_axis=-1, expand_axis=-1)
+    else:
+        return CombineInputs(combine_axis=-1, expand_axis=None)
+
+
+def _get_arch_layer(key, kwargs):
+
+    if key == "rnn-v1":
+        return RNN(**kwargs)
+    elif key == "rnn":
+        return HybridRNN(**kwargs)
+    elif key == "dense":
+        return MLPBlock(**kwargs)
+    elif key == "linear":
+        return MLPBlock(depth=0)
+    else:
+        raise KeyError(f"Unrecognized architecture provided: {key}")
+
+
+class _HiddenArchitecture(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        key: str,
+        net_kwargs: Mapping[str, Any],
+        feature_lengths: Mapping[str, int],
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        self.arch = _get_arch_layer(key, net_kwargs)
+        self.input_combiner = _get_combine_layer(key)
+        self.outputs = _get_output_layer(key, feature_lengths)
+
+    def call(self, tensors: Sequence[tf.Tensor]):
+
+        combined = self.input_combiner(tensors)
+        net_output = self.arch(combined)
+        return self.outputs(net_output)
+
+
+@dataclasses.dataclass
+class ArchitectureConfig:
```

Comment:
I removed all the special handling of input combining, the ML hidden network type, and output dispensing from `emulation.model._core` and the `MicrophysicsConfig`  and hid all of our architecture layers behind `ArcitectureConfig`.

Reply:
The `_HiddenArchitecture` with helper functions was simpler than creating a class for each model key, but that's an option if we wanted to make some of them public at some point in the future.

Reply:
I like the approach of hiding this logic behind `_HiddenArchitecture`. I think we are close enough for this PR, but probably need some further iteration/removing of layers.

Do you think a per-output-variable architecture configuration could be compatible with this API (e.g. RNN for temperature and Dense for humidity)? This potential use might break some of our assumptions.

Reply:
Hmm, yeah I didn't really design this for per-variable architectures and that would break the assumptions currently.  Not sure if the easiest way to try this would be to just have multiple `MicrophysicsConfigs` and some multi-model emulator, or push down I/O information to the architecture config.  I think as long as the per-variable architectures didn't need to share any information the former would be easy.

Reply:
> sure if the easiest way to try this would be to just have multiple MicrophysicsConfigs and some multi-model emulator

That probable makes the most sense...although maybe it's just cleaner to define new architectures if we want that kind of control.